### PR TITLE
Add roadmap

### LINKED ISSUE #10461: Roadmap Brainstorm
List 3 things that you think should be prioritised in the medium-long term

These are typically higher-level goals rather than individual features.

The purpose of this is to produce a document to communicate to contributors what certain things are more likely to be accepted, as they're wanted sooner. It's in no means a strict roadmap.

### PR DESCRIPTION
**Pull request by rubenwardy**
Closes #10461

After this PR is merged, I will add a link to it on the website and also rebase #8768.

### CODE REVIEW

**Comment by Zughy** [comment]
I know it may sound too harsh, but how about closing _new_ PRs that are not related to the roadmap or that are not a bugfix? I fear people will keep pushing whatever they want and the current problem is core devs have too many things to handle.

**Comment by paramat** [comment]
> but how about closing new PRs that are not related to the roadmap or that are not a bugfix?

That would obviously be completely impractical. All types of PR have to progress for MT to develop, even if there is a moderate prioritisation of the roadmap. That would mean no more mapgen PRs etc.

I was actually about to express my concern about the 'will be prioritised' line. It is dangerous in how it could be used by community, contributors or developers to try to halt necessary roadmap-unrelated PRs.
It would also be a lie, as we will obviously continue to process whatever PRs MT needs, many of which wil not be roadmap-related, even though we will raise the priority of roadmap-related PRs.

I think we should not state:
> Pull requests that address these goals will be prioritised.

This needs to be very carefully restated. It is too extreme and inflexible.

**Comment by rubenwardy** [comment]
will be prioritised is vague enough to allow us to be flexible

**Comment by paramat** [comment]
I consider it not vague, it is absolute: 'will' is absolute and 'prioritised' is absolute.
Keep in mind that due to lack of time to process everything, anything not prioritised is never done. 'Priority' essentially means 'what gets done'.
Even if we do not mean it that way, which we do not, the wording can and will be used against us.
I disapprove this PR with that line, it needs to be stated more moderately.

>  I fear people will keep pushing whatever they want because they want a certain feature

Do not worry, they will always do this whatever the roadmap is, it is not a problem because anything considered low priority or unnecessary will just be closed or left unreviewed.
Also, roadmap-unrelated contributions are essential for MT development and should continue.

**Comment by rubenwardy** [comment]
better?

**Comment by Zughy** [comment]
> Do not worry, they will always do this whatever the roadmap is, it is not a problem because anything considered low priority or unnecessary will just be closed or left unreviewed.

I think this is the real problem. Try to be one of those contributors who didn't notice the roadmap: how would you feel if no core dev didn't consider you for months (or years)? Wouldn't be better to just say "no" from the beginning? I don't want to bring discord in here, but there is a situation on the forum, and behaviours like these are exactly the ones feeding it

**Comment by Kezii** [comment]
Please take a look on how KDE handles the roadmap: 
https://pointieststick.com/2020/10/21/inside-kde-leadership-and-long-term-planning/
https://dot.kde.org/2019/09/07/kde-decides-three-new-challenges-wayland-consistency-and-apps

Project goals should not be too precise or in any way enforced, this is a community project (just like KDE) and the community contributes to what it wants. Issues and PRs outside the roadmap should not be de-prioritized in any way.

Project goals should be used to clarify what are the major pain points of the software, issues should be voted by users and stakeholders, goals should be used to explain said pain points in a way that gets everyone on the same page.



For what is happening right now on minetest, the following statement:

> "i should clarify to everyone that this roadmap will be decided by the core devs, not by democratic vote."

 doesn't make a lot of sense sense, if core devs want to impose a roadmap that is too different from the "democratic roadmap", then they better have a lot of free time because people are not going to follow that easily. And all of this become a big waste of time.

Without a true democratic vote and, an actual change in the development process (more core devs, less "not a bug" behavior from the core devs, ways to reduce workload on core devs, ways to reduce PR times), this is all feeling like a trick to get people to shut up and then go back to the old way of doing things, after the dust settles.

**Comment by rubenwardy** [comment]
For what it's worth, I disregarded that and used the overall tally instead

File: doc/direction.md
Comment (paramat) [inline doc/direction.md:8]: These are not really 'his opinions', the first 2 are actually MT's official roadmaps, just as valid as the priorities we have discussed.
Describing these as 'his opinions' seems to be undermine his authority and the official roadmaps.
Maybe you just worded this badly.

File: doc/direction.md
Comment (paramat) [inline doc/direction.md:5]: The documents linked in this 'Background' section are not really 'context for Minetest's direction', they are just as much part of Minetest's direction as our recently discussed priorities.
The wording 'To give context to this document' seems to be undermining his authority and the official roadmaps.

What is linked here in the 'Background' section should be described as other parts of the official roadmap.

File: doc/direction.md
Comment (paramat) [inline doc/direction.md:16]: This is not MT's roadmap or direction, only a part of it.

File: doc/direction.md
Comment (rubenwardy) [reply doc/direction.md:8]: Authority is proportional to time invested

File: doc/direction.md
Comment (paramat) [reply doc/direction.md:16]: Please can you explain why i am wrong? Happy to accept that if i am.

**Comment by paramat** [comment]
Zughy, i probably worded that quote very badly, sorry.

I do not imply we give no guidance on what contributions are preferred.
Discussions in issues already give a good indication.
We also encourage contributors to consider opening an issue to discuss a potential contribution and how desirable it is.
Once a contribution is made we discuss it and how desirable it is in the PR.
There is a lot of discussion and guidance.

What i meant was:
Despite all that happening, if a contributor insists on making a clearly undesirable contribution then no harm is done, because we will just disapprove it or close it.

I also agree that if a PR is unwanted but not disapproved for a period of time we should decide on a close instead of just letting it sit for years. I am a proponent for a stricter closing of issues and PRs for the sake of deveopers and contributors. I think we have been too soft to the point it actually harms everyone.

In discussions we already say 'no' to certain ideas, so we are already doing what you suggest.

File: doc/direction.md
Comment (paramat) [reply doc/direction.md:8]: Celeron55 has far more commits than anyone else, and possibly has done more work than anyone else too. But of course less than the total of everyone else.
But that is partially irrelevant because he is the original creator and occupies a special position.
I hope you do not mean it this way but this implies you are personally deciding to undermine him and personally deciding on the transfer of power. Has this been discussed with and agreed by other core devs? I have not seen any discussion.

Comment (rubenwardy) [reply doc/direction.md:8]: celeron55 delegated control of development to core developers when he appointed us and stopped actively contributing. His documents are useful in outlining high-level goals, but are outdated and aren't focused. A roadmap is a collection of goals and a plan to obtain these, this certainly matches that

**Comment by paramat** [comment]
Rubenwardy, thanks, it is ok now.

Is this WIP?
I noticed that many of the priorities listed by core devs are not included. This is obviously a problem because unless all are included this roadmap it is essentially being decided by you though selection, and i am sure you realise this is not right.

See my line comments. I am concerned by how this is worded in a way that seems to be undermining the other official roadmaps and celeron55. This is something i have seen you do before so it concerns me it is intentional.
Any transfer of authority has not been discussed with celeron55 and the other core devs.

This is not MT's primary roadmap or direction, as implied by the document, it is just one part alongside existing documents and no more valid than the other parts. This was not discussed and agreed, you have personally decided it.
> For what it's worth, I disregarded that and used the overall tally instead

Ah ok, i thought it was done on core dev tally.
I disagree with this approach, but i support core devs deciding to include priorities of non-core-devs if they are good ideas.

How do you have the authority to decide:
* The existing roadmaps are now less relevant.
* This is now the primary roadmap.
* Who decides the roadmap.
* Which small selection of priorities from those chosen people to include.

?
This has not been discussed.

I do not oppose having a document like this, but there is so much wrong about how this has been done.
I accept it is very difficult to make such a document.
:-1:

**Comment by rubenwardy** [comment]
I choose the top 4 goals of the core dev list, which also included the top goals in the general list

I will disregard your disapproval as it has very little value to me, given how much you use it

File: doc/direction.md
Comment (paramat) [reply doc/direction.md:8]: That makes some sense, but my concerns still stand about the wording.
We do not have absolute control of development, and this has not been discussed with him and core devs and decided.
His documents are more relevant than you imply, they are old but not outdated, they are the fundamental philosophy of MT.

**Comment by rubenwardy** [comment]
I can make it more clear that these are medium term goals, whereas the other resources are available as longer-term, guiding aims and philosophies

**Comment by kilbith** [comment]
Capture d’écran (2)
https://dev.minetest.net/Organisation

**Comment by benrob0329** [comment]
> 'Priority' essentially means 'what gets done'

Isn't this the point? If you spend time on whatever, then there is no point in making a roadmap to begin with as things will remain exactly the same.

-------

> How do you have the authority to decide

This is ridiculous and creating unnecessary discord. The Github is open and this is clearly meant for discussion just like every other PR here. Take issue with the contents, the initial suggestion is open for discussion for a reason.

> Frankly, your disapproval doesn't have too much value to me, you were always going to disapprove this like you do with everything else

This is also unnecessary and will only cause this to become more heated and controversial than it was already destined to be.

Please, don't make this a flame war resulting in absolutely no progress for the project.

**Comment by celeron55** [comment]
@paramat your exact concerns don't come across very clearly at all, but I would like to state that yes, core devs can set a roadmap to follow, and publish it officially, as often as they please. And yes, core devs can ignore the roadmaps of inactive core devs if they please. The roadmap doesn't have to follow any rules or prior roadmaps at all - that is left completely to the core team's discretion.

**Comment by Zughy** [comment]
As things have been addressed, I push again for declaring a duration to define what's "medium". KDE goals are biennial, Minetest is not as big as them though. How about including two releases, aka a year, to see how this go? Preceded by the same roadmap survey, like a month or two earlier. This could help not fossilising

**Comment by rubenwardy** [comment]
I'd say medium-term is 2 years, compared to long-term being the overall future of the project.
It's not feasible to get these goals done in less that 2 years, even 2 years is likely to be a stretch. Perhaps we could get the UI and entity goals done, but the others will be ongoing for quite a long time

**Comment by Zughy** [comment]
I see your point. I just want to specify people can vote for the same thing they voted the year before if they think it's not been addressed fully and they still believe it's a priority

**Comment by rubenwardy** [comment]
This should be reviewed yearly, even if the complete implementation of the goals would take longer.

There was a discussion with celeron55 and sfan5 on dealing with non-roadmapped pull requests.
TL;DR: pull requests that aren't on the roadmap should be closed unless they are concept-approved within a week. (Bugfixes should be exempt from this.)

Thoughts?

> **\<sfan5\>** rubenwardy: do you intend the roadmap document to be reassesed and edited once every \<some timeframe\>?
> **\<rubenwardy\>** yes
> **\<celeron55\>** it should be
> **\<rubenwardy\>** something like 2 years, or more often as needed
> **\<rubenwardy\>** depends when they become stale
> **\<rubenwardy\>** are you referring to issue links?
> **\<sfan5\>** for example those
> **\<celeron55\>** i would say a roadmap set by active devs should be completely re-prioritized every year
> **\<sfan5\>** but in general just when a goal is considered to be finished by devs it should disappear from the document
> **\<celeron55\>** oh you're assuming something will get done? 8)
> **\<sfan5\>** who knows
> **\<sfan5\>** my next question is what about issues and PRs that don't match the roadmap
> **\<sfan5\>** issues canbe accepted/discussed/rejected and be left sitting until they're relevant
> **\<rubenwardy\>** I don't think they should be rejected without thought, they can still be overall more positive
> **\<sfan5\>** but for PRs you'd want a clear "yes" "no" or the same frustrations as now will stay
> **\<celeron55\>** can we afford to try a stricter model for some time just to see what happens?
> **\<celeron55\>** i mean, that's what everyone seems to be asking for
> **\<rubenwardy\>** I'd say that being orthogonal to a roadmap is reason for closure, but there should be some grey areas
> **\<rubenwardy\>** bug fixes should obviously be allowed, and anything that's on a longer term road map as well
> **\<rubenwardy\>** grey areas -> room for human judgement
> **\<sfan5\>** entity improvements happen to be on the roadmap, but (if it wasn't) there are many points on my personal todo list that'd go in that direction
> **\<sfan5\>** and I wouldn't want them to be rejected if I was willing to work on them
> **\<sfan5\>** it's mostly small features, so the judgement could be based on that
> **\<sfan5\>** (otherwise you'd get users asking why coredevs have the privilege of ignoring the roadmap for their own desired features)
> **\<sfan5\>** anyway if the roadmap goes in what we really need is a shared queued of PRs to be considered
> **\<celeron55\>** my answer would be, core devs have to have extra privileges in exchange for having extra responsibilities
> **\<celeron55\>** otherwise nobody would want to be one
> **\<sfan5\>** hm, then users could convince coredevs to accept features they want but that aren't on the roadmap, sounds useful
> **\<celeron55\>** well the question will be, is there any practical difference to the current situation
> **\<celeron55\>** or
> **\<celeron55\>** what is the practical difference
> **\<sfan5\>** not really
> **\<sfan5\>** the difference would be that PRs would get a more or less clear rejection instead of lying around ignored
> **\<sfan5\>** and that the "you can convince coredevs of your feature" is more clearly communicated
> **\<celeron55\>** can we promise every PR will get a "review of objective" (what's a better wording for that?), which will always pass if the PR is included in the roadmap, and immediately a binding pass or fail otherwise
> **\<sfan5\>** I'd say yes
> **\<celeron55\>** or, well, not immediately but in a tight timeframe
> **\<celeron55\>** not sure how the voting process would work
> **\<celeron55\>** it needs to be light
> **\<sfan5\>** if this was a company you'd have a meeting to discuss
> **\<celeron55\>** make it so that any core dev can fail but two can pass, and the two that pass then handle the PR until it's merged (unless others want to help too)
> **\<celeron55\>** s/fail/close/
> **\<celeron55\>** if somebody closes a PR but then two coredevs decide to take it, they can then reopen it
> **\<sfan5\>** sounds good
> **\<celeron55\>** and they will be closed in a constant time, during which the pass can happen
> **\<celeron55\>** that time will have to be decided
> **\<celeron55\>** 3 days? a week?
> **\<celeron55\>** 3 hours? 8)
> **\<sfan5\>** a week would be a good minimum to avoid the risk that devs don't have time to look at issues once
> **\<celeron55\>** so a PR marked as "Not on roadmap" will be closed after a week once a triager or a core dev notices it's still there
> **\<celeron55\>** if the 2-dev special pass happens, what would that label be
> **\<celeron55\>** i guess "Object accepted" or something
> **\<celeron55\>** "Objective accepted"*
> **\<celeron55\>** i'd try to not sound very bureaucratic
> **\<celeron55\>** even while we are
> **\<celeron55\>** then just scroll down the PRs until they show as one week old and close the ones that have neither of those labels
> **\<celeron55\>** umm
> **\<celeron55\>** i mean, have the former and don't have the latter
> **\<celeron55\>** "Accepted but not on roadmap" would be a very clear label and you wouldn't even need to read any rules to understand what it means
> **\<celeron55\>** whatever
> **\<rubenwardy\>** should "Supported by core dev" on issues be used to pre-approve things for "Accepted but not on roadmap"?
> **\<sfan5\>** should objective include the design of the PR?
> **\<sfan5\>** someone could implement the right feature but go the wrong way about it (would be with a glance at the code)
> **\<sfan5\>** +obvious
> **\<celeron55\>** if it's not on the roadmap, then just don't pass it
> **\<celeron55\>** if it is on the roadmap, then it some kind of guidance of doing it properly should be written and then closed
> **\<celeron55\>** -it
> **\<celeron55\>** rubenwardy: i think no, it's for the merge process
> **\<celeron55\>** or umm
> **\<celeron55\>** i mean, we'll have two things that require two core devs, it will become a mess
> **\<celeron55\>** i'd just keep the pass discussion in the comments if possible. of course if it becomes a mess too then decide on the lesser mess

**Comment by v-rob** [comment]
So, it appears that there will be three new tags: "Not on roadmap", "Objective accepted", and "Accepted but not on roadmap", or something along that wording, along with the tag "Bugfix".  I like it, it's very clear and shouldn't arouse confusion.  So then, what will happen with existing PRs?  Will they be exempt from the rule, or do we go through all of them and apply this system to every one?

As for issues, I assume the same idea applies except "Bugfix" is replaced with "Bug".  But going through all the existing issues would take even longer than the PRs, even if the "Bug"s and "Unconfirmed bug"s are filtered out.  (And what do we do about "Unconfirmed bug"s?  Sometimes they're left sitting for [months or years](https://github.com/minetest/minetest/issues?q=is%3Aopen+is%3Aissue+label%3A%22Unconfirmed+bug%22+sort%3Acreated-asc).)

**Comment by paramat** [comment]
Thanks, this is presented better now.

My remaining requests/suggestions for this are:

* Wait longer (2+ weeks?) for input in the issue from pyrollo, and maybe input from the long-inactive core devs (but obviously do not wait any longer for them).
For this reason i think this PR is WIP.
* Then, include all the suggested priorities of all the core devs who have stated them, with less assumption of one being contained in another.
For example, my priority of making the engine game-neutral and removing MTG has not been included. Although related, this is not implied by 'new main menu'.
There are probably other suggested priorities that have been omitted by a too-aggressive compression into fewer stated priorities.
It is better to error on the side of stating too many than too few.
I think it is fine to add priorities from non core devs too, as long as these are agreed by at least 2 core devs, not by an individual.
* Although 2 core devs suggested fancy graphics/shaders improvements, this document decides these are lower priority and to be done later. This does not seem right, priorities are priorities (even though i agree with you).

**Comment by rubenwardy** [comment]
> Wait longer (2+ weeks?) for input in the issue from pyrollo, and maybe input from the long-inactive core devs (but obviously do not wait any longer for them).

I agree

>  For example, my priority of making the engine game-neutral and removing MTG has not been included. Although related, this is not implied by 'new main menu'.

Yes, I folded this into UI improvements as I felt this was related. The main task with removing MTG is I improving the mainmenu to show multiple games. I've explicitly put this

> Although 2 core devs suggested fancy graphics/shaders improvements, this document decides these are lower priority and to be done later. This does not seem right, priorities are priorities (even though i agree with you).

True, I felt like it was important to make "repair the unstable foundations before building on them" clear in the description. Thoughts, @lhofhansl and @nerzhul?

**Comment by Zughy** [comment]
#10057 should be closed too if/when this gets merged

**Comment by GenshinMT** [comment]
Regarding to roadmap, core devs should mainly prioritize more on performance issues and bug fixes first before working on new features based on road map to ensure clean feature releases.

**Comment by Zughy** [comment]
@GenshinMT as I'm the one cleaning the code in these days, I can assure you the "code rot" point is spot on. Whether we're talking about features, performance issues or bug fixes, it's way harder to do any of those without a clean space

about v-rob PR doubt: how about keeping the current ones even if they don't match the roadmap, as the creators couldn't foresee the future? And to solve them on the go. Theoretically, in doing so, PR amount should decrease with the passing of time anyway *while* those contributors didn't waste their time and energies

**Comment by GenshinMT** [comment]
@Zughy I agree, cleanup is always the first step before working on patches and performance issues.
_"No sane person would dare work on things with a messy plate on the table"_

**Comment by Zughy** [comment]
More than a month has passed and every core dev spoke their mind, so: can this be merged and considered for 5.4.0+ ?

**Comment by v-rob** [comment]
I have no objections to the PR itself, but I would also like the new tags for issues/PRs to be formalized as well.

**Comment by Zughy** [comment]
That, and what is going to happen to old PRs not aligning with the roadmap

**Comment by v-rob** [comment]
Under "UI Improvements", can something be added directing people to consult me before making large UI changes?  I'd like to ensure that work in this area remains coordinated so that there aren't conflicts.

**Review by v-rob — APPROVED** [review]
Looks good to me.  Can there be other core dev approvals?  We can figure tagging/whatever later, say, in a meeting, without it being in this PR.

**Comment by rubenwardy** [comment]
I think that all active core developers should consider this, rather than the usual 2 approvals

**Comment by v-rob** [comment]
Yeah, that's what I meant (otherwise I wouldn't have asked for more approvals seeing as you and I are both core devs).

**Review by SmallJoker — APPROVED** [review]
The 1 week deadline for concept approvals is quite short, but I think that it's not a strict rule after all.
Label either as `Concept approved` or `Roadmap` - that's a good idea.
Good points in the overall direction and goals. No objections.

**Review by sfan5 — APPROVED** [review]

**Comment by sofar** [comment]
:+1: from me

**Pull request merged**
